_posts/2019/2019-05-11-how-to-improve-your-components-with-react-hooks.md

```diff
@@ -0,0 +1,182 @@
+---
+layout: post
+title: How To Improve Your Components With React Hooks
+date: 2019-05-11
+header_image: public/react-hooks.png
+header_position: center
+category: coding
+tags: ["javascript", "reactjs"]
+authors: ["Jonathan"]
+about_authors: ["jwieben"]
+---
+
+## How To Improve Your Components With React Hooks
+
+The React world has been going crazy over the new Hooks API that was released a few months ago. New libraries for hooks are appearing left and right and everyone seems to be rewriting their apps with them.
+
+So what is it about hooks that people are so excited about? Well, there are a variety of reasons. In this post I would like to focus on one of the central advantages that hooks bring to the table: **They allow you to organize your component logic in a simpler way.**
+
+I am going to illustrate this by writing the same component in two ways, with and without hooks, and later compare them.
+
+#### A Typical Class Component
+
+Let's assume we wanted to write a `UserProfile` component, that displays a users profile picture next to their name. We also want the profile picture to be hidden when the window width is fewer than 1024 pixels.
+
+In order to achieve this, we are going to need some state. It should store the user data as well as a flag indicating if we are on a small screen. We will use the components lifecycle methods to set and update these values.
+
+Here's how such a component might look:
+
+```js
+class UserProfile extends React.Component {
+  state = {
+    isMobile: window.innerWidth <= 1024,
+    user: null
+  }
+
+  componentDidMount() {
+    this.fetchUser()
+    window.addEventListener('resize', this.handleResize)
+  }
+
+  componetDidUpdate(prevProps) {
+    if (this.props.userId !== prevProps.userId) {
+      this.fetchUser()
+    }
+  }
+
+  componentWillUnmount() {
+    window.removeEventListener('resize', this.handleResize)
+  }
+
+  handleResize = () => {
+    if (this.state.isMobile && window.innerWidth > 1024) {
+      this.setState({ isMobile: false })
+    } else if (!this.state.isMobile && window.innerWidth <= 1024) {
+      this.setState({ isMobile: true })
+    }
+  }
+
+  fetchUser = () => {
+    fetch('https://some-api.com/user/' + this.props.userId)
+      .then(user => this.setState({ user }))
```

Comment:
`fetch().then()` gives you a response object `response` with methods like `response.json()`. `response.json()` returns again a promise that resolves with the value.
```suggestion
      .then(response => response.json())
      .then(user => this.setState({ user }))
```